scripts/update_lit_checks.py

```diff
@@ -0,0 +1,171 @@
+#!/usr/bin/env python3
+# Copyright 2021 WebAssembly Community Group participants
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A test case update script.
+
+This script is a utility to update wasm-opt based lit tests with new FileCheck
+patterns. It is based on LLVM's update_llc_test_checks.py script.
+"""
+
+import argparse
+import glob
+import os
+import re
+import subprocess
+import sys
+import tempfile
+
+
+script_name = os.path.basename(__file__)
+NOTICE = f';; NOTE: Assertions have been generated by {script_name}'
+RUN_LINE_RE = re.compile(r'^\s*;;\s*RUN:\s*(.*)$')
+CHECK_PREFIX_RE = re.compile(r'.*--check-prefix[= ](\S+).*')
+FUNC_RE = re.compile(r'^\s*\(func \$(\S*).*$', re.MULTILINE)
+
+
+def warn(msg):
+    print(f'WARNING: {msg}', file=sys.stderr)
+
+
+def itertests(args):
+    'Yield (filename, lines) for each test specified in the command line args'
+    for pattern in args.tests:
+        tests = glob.glob(pattern)
+        if not tests:
+            warn(f'No tests matched {pattern}. Ignoring it.')
+            continue
+        for test in tests:
+            with open(test) as f:
+                lines = [line.rstrip() for line in f]
+            first_line = lines[0] if lines else ''
+            if script_name not in first_line and not args.force:
+                warn(f'Skipping test {test} which was not generated by '
+                     f'{script_name}. Use -f to override.')
+                continue
+            yield test, lines
+
+
+def find_run_lines(test, lines):
+    matches = [m.group(1)
+               for m in [RUN_LINE_RE.match(l) for l in lines]
+               if m]
+    if not matches:
+        warn(f'No RUN lines found in {test}. Ignoring.')
+        return
+    run_lines = [matches[0]]
+    for line in matches[1:]:
+        if run_lines[-1].endswith('\\'):
+            run_lines[-1] = run_lines[-1].rstrip('\\') + ' ' + line
+        else:
+            run_lines.append(line)
+    return run_lines
+
+
+def run_command(args, test, tmp, command):
+    env = dict(os.environ)
+    env['PATH'] = args.binaryen_bin + os.pathsep + env['PATH']
+    command = command.replace('%s', test)
+    command = command.replace('%t', tmp)
```

Comment:
Are these two substitutions really all the RUN lines can do?  Won't be be playing catchup with lit here are we re-implement the RUN -line mechanism here?

I guess the other thing RUN lines do is substitute the short command for the full command name, and you have chosen instead to inject the a PATH element..

Reply:
No, the full list of substitutions is quite extensive (and extensible): https://llvm.org/docs/CommandGuide/lit.html#substitutions. Besides the tool names, which I covered by modifying PATH, these are the only substitutions I use regularly in tests, though. It shouldn't be too onerous to write any test that requires more esoteric substitutions by hand rather than using this script.